plugin/healthchecker/checkers/icmp.go

```diff
@@ -0,0 +1,216 @@
+package checkers
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"time"
+
+	"github.com/coredns/caddy"
+	"github.com/coredns/coredns/plugin/pkg/log"
+	"golang.org/x/net/icmp"
+	"golang.org/x/net/ipv4"
+	"golang.org/x/net/ipv6"
+)
+
+type (
+	ICMPChecker struct {
+		logger       log.P
+		isPrivileged bool
+		timeout      time.Duration
+	}
+
+	ICMPCheckerParams struct {
+		IsPrivileged bool
+		Timeout      time.Duration
+	}
+
+	icmpParams struct {
+		Network       string
+		ListenAddress string
+		ReplyType     icmp.Type
+		Proto         int
+		Msg           []byte
+	}
+)
+
+func ParseICMPParams(c *caddy.Controller) (*ICMPCheckerParams, error) {
+	prm := &ICMPCheckerParams{}
+
+	for c.NextBlock() {
+		switch c.Val() {
+		case "privileged":
+			args := c.RemainingArgs()
+			if len(args) != 0 {
+				return nil, fmt.Errorf("expected zero value param, but got '%v'", args)
+			}
+			prm.IsPrivileged = true
+		case "timeout":
+			args := c.RemainingArgs()
+			if len(args) != 1 {
+				return nil, fmt.Errorf("expected one value param, but got '%v'", args)
+			}
+			value := args[0]
+			timeout, err := time.ParseDuration(value)
+			if err != nil || timeout <= 0 {
+				return nil, fmt.Errorf("invalid timeout '%s'", value)
+			}
+			prm.Timeout = timeout
+		default:
+			return nil, fmt.Errorf("unknow ICMP parameter: '%s'", c.Val())
+		}
+	}
+
+	return prm, nil
+}
+
+// NewICMPChecker creates icmp checker.
+func NewICMPChecker(logger log.P, prm *ICMPCheckerParams) (*ICMPChecker, error) {
+	if prm.Timeout <= 0 {
+		prm.Timeout = defaultHTTPTimeout
+	}
+
+	if prm.IsPrivileged {
+		logger.Warningf("you run icmp checker in privileged mode, make sure Coredns has appropriate permission")
+	}
+
+	return &ICMPChecker{
+		logger:       logger,
+		isPrivileged: prm.IsPrivileged,
+		timeout:      prm.Timeout,
+	}, nil
+}
+
+func (c ICMPChecker) Check(endpoint string) bool {
+	isV4 := isIPv4(endpoint)
+	ip := net.ParseIP(endpoint)
+	if ip == nil {
```

Comment:
We can check if ip is v4 simpler: 
```
ip := net.ParseIP(endpoint)
isV4 := ip.To4() != nil
```

Reply:
Ok, thanks. 
But I think it should be written in comments (link or explanation) because it's not obvious why you chose the current implementation of check.